### PR TITLE
'tidy up' PR to improve the ease of calling submit_to_azure_if_needed.

### LINKED ISSUE #23: Simplify handling of core arguments to submit_if_needed.
* Automatically find `config.json` and `environment.yml` in current folder or any of the parent folders (stop going up when going out of a folder that is in PythonPath)

### LINKED ISSUE #27: Switch tests to use a separate AzureML workspace
* Create a new AzureML workspace and storage account only for the hi-ml package. Ideally, this should be tried out with Azure CLI tools - it would be ideal if we had a completely script-driven way of creating 
  * resource group 
  * workspace
  * storage account
  * service principal
  * data store
  * compute cluster
* Document all the necessary steps in an MD file, store the script(s) to create everything
* Switch the github workflow to this new workspace and service principal

### LINKED ISSUE #53: Ensure the submit_if_needed copes with str for all files and paths
All arguments that point to files are presently accepting Path objects. However, many people still use strings for handling file names. Ensure that all arguments are happily accepting PathOrString.

### PR DESCRIPTION
**Pull request by dumbledad**
Closes #53 
Closes #23 
Closes #27 

A 'tidy up' PR to improve the ease of calling `submit_to_azure_if_needed`.

Also taking the opportunity to swap our pipelines to use our own AzureML workspace.

Note that #94 is closed by adding a rule to the datastore deleteing blobs matching the input or output test prefix pattern after 24 hours.

### CODE REVIEW

File: testhiml/health/azure/test_himl.py [hunk 1]
```diff
@@ -564,12 +603,17 @@ def test_invoking_hello_world_no_config(run_target: RunTarget, tmp_path: Path) -
         'body': 'print(f"The message was: {args.message}")'
     }
     extra_args = [f"--message={message_guid}"]
-    output = render_and_run_test_script(tmp_path, run_target, extra_options, extra_args, run_target == RunTarget.LOCAL)
     expected_output = f"The message was: {message_guid}"
     if run_target == RunTarget.LOCAL:
+        output = render_and_run_test_script(tmp_path, run_target, extra_options, extra_args,
+                                            run_target == RunTarget.LOCAL,
+                                            suppress_config_creation=run_target == RunTarget.AZUREML)
```
Comment (JonathanTripp) [inline testhiml/health/azure/test_himl.py:610]: suppress_config_creation is always False here, and that's the default, so maybe remove this line?

File: testhiml/health/azure/test_himl.py [hunk 2]
```diff
@@ -564,12 +603,17 @@ def test_invoking_hello_world_no_config(run_target: RunTarget, tmp_path: Path) -
...
         assert expected_output in output
     else:
-        assert "Cannot glean workspace config from parameters, and so not submitting to AzureML" in output
+        with pytest.raises(ValueError) as e:
+            render_and_run_test_script(tmp_path, run_target, extra_options, extra_args, run_target == RunTarget.LOCAL,
+                                       suppress_config_creation=run_target == RunTarget.AZUREML)
```
Comment (JonathanTripp) [inline testhiml/health/azure/test_himl.py:615]: And here: suppress_config_creation is True. Is it clearer if explicitly set to True?

Comment (dumbledad) [reply testhiml/health/azure/test_himl.py]: Done

File: testhiml/health/azure/test_himl.py (hunk 1 above)
Comment (dumbledad) [reply testhiml/health/azure/test_himl.py]: Done

File: src/health/azure/himl.py [hunk 3]
```diff
@@ -426,6 +441,37 @@ def submit_to_azure_if_needed(  # type: ignore
...
+            pythonpaths: List[Path]) -> Optional[Path]:
+        for child in start_at.iterdir():
+            if child.is_file() and child.name == file_name:
+                return child
+        if start_at.parent == start_at or start_at in pythonpaths:
+            return None
+        return return_file_or_parent(start_at.parent, file_name, stop_at_pythonpath, pythonpaths)
+
+    pythonpaths: List[Path] = []
+    if 'PYTHONPATH' in os.environ:
```
Comment (ant0nsc) [inline src/health/azure/himl.py:466]: why not use sys.path?

File: src/health/azure/himl.py [hunk 4]
```diff
@@ -426,6 +441,37 @@ def submit_to_azure_if_needed(  # type: ignore
     exit(0)
 
 
+def _find_file(file_name: str, stop_at_pythonpath: bool = True) -> Optional[Path]:
```
Comment (ant0nsc) [inline src/health/azure/himl.py:444]: do we ever NOT want to stop at pythonpath?

File: testhiml/health/azure/test_himl.py [hunk 5]
```diff
@@ -35,6 +35,7 @@
 
 INEXPENSIVE_TESTING_CLUSTER_NAME = "lite-testing-ds2"
 EXPECTED_QUEUED = "This command will be run in AzureML:"
+GITHUB_SHIBBOLETH = "GITHUB_RUN_ID"  # https://docs.github.com/en/actions/reference/environment-variables
```
Comment (ant0nsc) [inline testhiml/health/azure/test_himl.py:38]: what's a shibboleth? maybe keep the language simpler for the foreigners like me ;-)

**Review by ant0nsc — APPROVED** [review]

File: src/health/azure/download_aml_run.py [hunk 6]
```diff
@@ -27,7 +27,7 @@ def determine_output_dir_name(args: Namespace, run_id_source: AzureRunIdSource,
         output_path = output_dir / Path(args.latest_run_path).stem
     elif run_id_source == AzureRunIdSource.RUN_RECOVERY_ID:
         output_path = output_dir / args.run_recovery_ids.replace(":", "")
-    elif run_id_source == AzureRunIdSource.RUN_ID:
+    else:  # run_id_source == AzureRunIdSource.RUN_ID:
```
[the remark this answers is not in the document]
Comment (dumbledad) [reply src/health/azure/download_aml_run.py:30]: Can do, but I thought the comment was informative about the reamining option in the enum?

File: src/health/azure/himl.py (hunk 4 above)
Comment (dumbledad) [reply src/health/azure/himl.py:444]: People may have entries in PYTHONPATH that reach into a sub directory of their repository, no?

File: testhiml/health/azure/test_himl.py (hunk 5 above)
Comment (dumbledad) [reply testhiml/health/azure/test_himl.py:38]: Will do. It's like a 'tell' in poker, something that gives away something about yourself unwittingly.  It is derived from a scary bible verse: [Judges 12:6](https://www.biblegateway.com/passage/?search=Judges%2012%3A6&version=KJV).  If Google Translate is to be trusted the equivalent German word is erkennungszeichen.

File: src/health/azure/himl.py (hunk 3 above)
Comment (dumbledad) [reply src/health/azure/himl.py:466]: I can do, but they are not quite the same:

> A list of strings that specifies the search path for modules. Initialized from the environment variable PYTHONPATH, plus an installation-dependent default.

(From [here](https://docs.python.org/3/library/sys.html#sys.path).)

File: src/health/azure/himl.py [hunk 7]
```diff
@@ -379,6 +383,17 @@ def submit_to_azure_if_needed(  # type: ignore
         logging.info(f"No snapshot root directory given. Uploading all files in the current directory {Path.cwd()}")
         snapshot_root_directory = Path.cwd()
 
+    if workspace_config_path is None:
+        workspace_config_path = _find_file(WORKSPACE_CONFIG_JSON)
+        if workspace_config_path:
+            logging.info(f"Using the workspace config path found at {str(workspace_config_path.absolute())}")
```
[the remark this answers is not in the document]
Comment (dumbledad) [reply src/health/azure/himl.py:389]: Fixed

File: src/health/azure/download_aml_run.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ant0nsc) [reply src/health/azure/download_aml_run.py:30]: ah, OK, it very much looked like an accidental piece of code.

**Pull request merged**